service/history/api/updateworkflow/api.go

```diff
@@ -61,177 +61,204 @@ const (
+		// only if there were buffered events and because there were no pending WT, there can't be buffered events.
+		return nil, consts.ErrWorkflowTaskStateInconsistent
+	}
 
-			taskQueue = common.CloneProto(newWorkflowTask.TaskQueue)
-			normalTaskQueueName = ms.GetExecutionInfo().TaskQueue
-			directive = worker_versioning.MakeDirectiveForWorkflowTask(
-				ms.GetWorkerVersionStamp(),
-				ms.GetLastWorkflowTaskStartedEventID(),
-			)
-
-			return &api.UpdateWorkflowAction{
-				Noop:               true,
-				CreateWorkflowTask: false,
-			}, nil
-		},
-		nil,
-		shardCtx,
-		workflowConsistencyChecker,
+	u.scheduledEventID = newWorkflowTask.ScheduledEventID
+	if _, scheduleToStartTimeoutPtr := ms.TaskQueueScheduleToStartTimeout(ms.CurrentTaskQueue().Name); scheduleToStartTimeoutPtr != nil {
+		u.scheduleToStartTimeout = scheduleToStartTimeoutPtr.AsDuration()
+	}
+
+	u.taskQueue = common.CloneProto(newWorkflowTask.TaskQueue)
+	u.normalTaskQueueName = ms.GetExecutionInfo().TaskQueue
+	u.directive = worker_versioning.MakeDirectiveForWorkflowTask(
+		ms.GetWorkerVersionStamp(),
+		ms.GetLastWorkflowTaskStartedEventID(),
 	)
 
-	// Wrapping workflow context related operation in separate func to prevent usage of its fields
-	// (including any mutable state fields) outside of this func after workflow lock is released.
-	// It is important to release workflow lock before calling matching.
-	if err != nil {
-		// If update is received while WFT is running, it will be waiting for the next WFT.
-		// And if that running WFT completes workflow, then update is rejected (see CancelIncomplete).
-		// Special handling for consts.ErrWorkflowCompleted here is needed to keep parity with this.
-		// I.e. if update is received and workflow was completed, or is about to be completed,
-		// then update is consistently rejected (instead of returning error in some cases).
-		if errors.Is(err, consts.ErrWorkflowCompleted) {
-			rejectionResp := createResponse(
-				wfKey,
-				req,
-				&updatepb.Outcome{
-					Value: &updatepb.Outcome_Failure{Failure: update.CancelReasonWorkflowCompleted.RejectionFailure()},
-				},
-				enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED)
-			return rejectionResp, err
-		}
-		return nil, err
+	return &api.UpdateWorkflowAction{
+		Noop:               true,
+		CreateWorkflowTask: false,
+	}, nil
+}
+
+func (u *Updater) OnError(
+	err error,
+) (*historyservice.UpdateWorkflowExecutionResponse, error) {
+	// If update is received while WFT is running, it will be waiting for the next WFT.
+	// And if that running WFT completes workflow, then update is rejected (see CancelIncomplete).
+	// Special handling for consts.ErrWorkflowCompleted here is needed to keep parity with this.
+	// I.e. if update is received and workflow was completed, or is about to be completed,
+	// then update is consistently rejected (instead of returning error in some cases).
```

Comment:
I need to rephrase this. I wrote it and had to read 2 times to get what it is about.

Reply:
```suggestion
	// Special handling for consts.ErrWorkflowCompleted here is needed for consistency with the case when update is received while WFT is running and this WFT completes workflow. In this case update is rejected (see update.CancelIncomplete).
```